src/mito2/src/sst/index/fulltext_index/applier.rs

```diff
@@ -187,6 +217,188 @@ impl FulltextIndexApplier {
+        row_groups: impl Iterator<Item = (usize, bool)>,
+    ) -> Result<Option<Vec<(usize, Vec<Range<usize>>)>>> {
+        let timer = INDEX_APPLY_ELAPSED
+            .with_label_values(&[TYPE_FULLTEXT_INDEX])
+            .start_timer();
+
+        let (input, mut output) = Self::init_coarse_output(row_groups);
+        let mut applied = false;
+
+        for (column_id, request) in &self.requests {
+            if request.terms.is_empty() {
+                // only apply terms
+                continue;
+            }
+
+            applied |= self
+                .apply_coarse_one_column(
+                    file_id,
+                    file_size_hint,
+                    *column_id,
+                    &request.terms,
+                    &mut output,
+                )
+                .await?;
+        }
+
+        if !applied {
+            timer.stop_and_discard();
+            return Ok(None);
+        }
+
+        Self::adjust_coarse_output(input, &mut output);
+        Ok(Some(output))
+    }
+
+    async fn apply_coarse_one_column(
+        &self,
+        file_id: FileId,
+        file_size_hint: Option<u64>,
+        column_id: ColumnId,
+        terms: &[FulltextTerm],
+        output: &mut [(usize, Vec<Range<usize>>)],
+    ) -> Result<bool> {
+        let blob_key = format!("{INDEX_BLOB_TYPE_BLOOM}-{column_id}");
+        let Some(reader) = self
+            .index_source
+            .blob(file_id, &blob_key, file_size_hint)
+            .await?
+        else {
+            return Ok(false);
+        };
+        let config =
+            Config::from_blob_metadata(reader.metadata()).context(ApplyFulltextIndexSnafu)?;
+
+        let predicates = Self::terms_to_predicates(terms, &config);
+        if predicates.is_empty() {
+            return Ok(false);
+        }
+
+        let range_reader = reader.reader().await.context(PuffinBuildReaderSnafu)?;
+        let reader = if let Some(bloom_filter_cache) = &self.bloom_filter_index_cache {
+            let blob_size = range_reader
+                .metadata()
+                .await
+                .context(MetadataSnafu)?
+                .content_length;
+            let reader = CachedBloomFilterIndexBlobReader::new(
+                file_id,
+                column_id,
+                Tag::Fulltext,
+                blob_size,
+                BloomFilterReaderImpl::new(range_reader),
+                bloom_filter_cache.clone(),
+            );
+            Box::new(reader) as _
+        } else {
+            Box::new(BloomFilterReaderImpl::new(range_reader)) as _
+        };
+
+        let mut applier = BloomFilterApplier::new(reader)
+            .await
+            .context(ApplyBloomFilterIndexSnafu)?;
+        for (_, output) in output.iter_mut() {
+            // All rows are filtered out, skip the search
+            if output.is_empty() {
+                continue;
+            }
+
+            *output = applier
+                .search(&predicates, output)
```

Comment:
non-blocking: Can we rename the loop variable `output`? It's a little confusing.
```suggestion
        for (_, row_group_output) in output.iter_mut() {
            // All rows are filtered out, skip the search
            if row_group_output.is_empty() {
                continue;
            }

            *row_group_output = applier
                .search(&predicates, row_group_output)
```